torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -858,7 +876,20 @@ def _reduce_grad(state: _FSDPState, handle: FlatParamHandle) -> torch.Tensor:
             group=state.process_group,
         )
         if uses_hybrid_sharded_strategy:
-            dist.all_reduce(new_sharded_grad, group=state._inter_node_pg)
+            state._all_reduce_stream.wait_stream(state._post_backward_stream)
```

Comment:
This captures the data dependency between reduce-scatter and all-reduce.

The reduce-scatter happens in the `_post_backward_stream`, so having this `_all_reduce_stream` wait on it ensure that the all-reduce uses the reduce-scattered data.

cc: @ezyang